tools/istio-iptables/pkg/dependencies/implementation.go

```diff
@@ -61,13 +63,60 @@ var XTablesCmds = sets.New(
 	constants.IP6TABLESSAVE,
 )
 
+// XTablesWriteCmds contains all xtables commands that do write actions (and thus need a lock)
+var XTablesWriteCmds = sets.New(
+	constants.IPTABLES,
+	constants.IP6TABLES,
+	constants.IPTABLESRESTORE,
+	constants.IP6TABLESRESTORE,
+)
+
 // RealDependencies implementation of interface Dependencies, which is used in production
 type RealDependencies struct {
+	IptablesVersion  IptablesVersion
 	NetworkNamespace string
-	HostNSEnterExec  bool
 	CNIMode          bool
 }
 
+const iptablesVersionPattern = `v([0-9]+(\.[0-9]+)+)`
+
+type IptablesVersion struct {
+	// the actual version
+	version *utilversion.Version
+	// true if legacy mode, false if nf_tables
+	legacy bool
+}
+
+// NoLocks returns true if this version does not use or support locks
+func (v IptablesVersion) NoLocks() bool {
+	// nf_tables does not use locks
+	// legacy added locks in 1.6.2
+	return !v.legacy || v.version.LessThan(IptablesRestoreLocking)
+}
+
+func DetectIptablesVersion(ver string) (IptablesVersion, error) {
+	if ver == "" {
+		var err error
+		verb, err := exec.Command("iptables", "--version").CombinedOutput()
+		if err != nil {
+			return IptablesVersion{}, err
+		}
+		ver = string(verb)
+	}
+	// Legacy will have no marking or 'legacy', so just look for nf_tables
```

Comment:
you may add a comment with examples of the output

```
$ iptables --version
iptables v1.8.9 (nf_tables)
$ iptables-legacy --version
iptables v1.8.9 (legacy)
```

for paranoid mode just follow the symlinks

```
$ which iptables
/usr/sbin/iptables
aojea@aojea:~$ readlink /usr/sbin/iptables
/etc/alternatives/iptables
aojea@aojea:~$ readlink /etc/alternatives/iptables
/usr/sbin/iptables-nft
```

Reply:
nevermind, is documented on the unit test tools/istio-iptables/pkg/dependencies/implementation_test.go